pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"google.golang.org/grpc"
+)
+
+// PluginInfo contains gRPC connection info
+type PluginInfo struct {
+	SockPath string
+}
+
+// WatchPlugins watches for the creation of plugin sockets at the path
+func WatchPlugins(path string, outChan chan PluginInfo) error {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return err
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	err = watcher.Add(path)
+	if err != nil {
+		watcher.Close()
+		return err
+	}
+
+	go func() {
+		for _, f := range files {
+			outChan <- PluginInfo{path + f.Name()}
+		}
+		for {
+			event := <-watcher.Events
+			if event.Op&fsnotify.Create == fsnotify.Create {
+				outChan <- PluginInfo{event.Name}
+			}
+		}
+	}()
+	return nil
+}
+
+// Dial establishes the gRPC communication with the picked up plugin socket. https://godoc.org/google.golang.org/grpc#Dial
+func Dial(unixSocketPath string) (*grpc.ClientConn, error) {
+	c, err := grpc.Dial(unixSocketPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(10*time.Second),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// GetResourceNameFrom returns base resource name from the plugin socket absolute path
+func GetResourceNameFrom(sockPath string) (string, error) {
+	if !((strings.Contains(sockPath, ".sock")) && (strings.Contains(sockPath, "/"))) {
+		return "", fmt.Errorf("invalid sock name: %s. Expected format: <sock-dir>/<domain>.<resource>.sock", sockPath)
```

Comment:
I thought it is meant to be built as a common mechanism for both device-plugin and other potential components of kubelet.
So, it’s specially for device-plugin indeed?

Reply:
As I was mentioning higher, we shouldn't impose a name format for the sockets. We're trying to fit a DNS name into an FS naming convention, that's not a good model (as shown by the further transformation we have to do on the string) and is prone to implementation errors and user errors.

Reply:
This pkg is to be shared by other components, like CSI, as well. +1 to avoid specific naming conventions.

@saad-ali @verult could you comment on any socket naming requirements/preference from CSI or flexvolume side?

Reply:
FYI, according to CSI proposal, the expected naming convention is /var/lib/kubelet/plugins/[SanitizedCSIDriverName]/csi.sock

For simplicity, wonder whether we can have csi plugins go to /var/lib/kubelet/plugins/csi/ directory and device plugins go to /var/lib/kubelet/plugins/device_plugins/? The parent directory can be a parameter passed in to NewPluginWatcher. Perhaps we can just pass the absolute path in the channel, and the PluginWatcher caller can do its own naming convention verification?

Reply:
Can you point me to that convention ? I can't seem to find any directive in [the spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#configuration-and-operation).

Reply:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md

Reply:
Flexvolume doesn't use sockets to communicate with kubelet; instead, kubelet executes Flexvolume drivers directly on host. I don't think this model works well for Flex, and IMO that's OK.

Reply:
> For simplicity, wonder whether we can have csi plugins go to /var/lib/kubelet/plugins/csi/ directory and device plugins go to /var/lib/kubelet/plugins/device_plugins/? 

Hmm, we mwntioned moving the checkpointing data out of this directory, I think that's a good idea. In that case where would we put it?
Wouldn't it make more sense to have:
- var/lib/kubelet/csi/
- var/lib/kubelet/device-plugin/
- var/lib/kubelet/device-plugin/data.checkpoint
- var/lib/kubelet/device-plugin/plugins/